backend/azure/dtmb/internal/utils/typeconversion.go

```diff
@@ -0,0 +1,315 @@
+package utils
+
+import (
+	"fmt"
+
+	"github.com/microsoft/durabletask-go/internal/protos"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+
+	dtmbprotos "github.com/microsoft/durabletask-go/backend/azure/dtmb/internal/backend/v1"
+)
+
+func ConvertFailureDetails(failureDetails *dtmbprotos.FailureDetails) *protos.TaskFailureDetails {
+	if failureDetails == nil {
+		return nil
+	}
+
+	return &protos.TaskFailureDetails{
+		ErrorType:      failureDetails.GetErrorType(),
+		ErrorMessage:   failureDetails.GetErrorMessage(),
+		StackTrace:     &wrapperspb.StringValue{Value: failureDetails.GetStackTrace()}, // what if this is not string
+		InnerFailure:   ConvertFailureDetails(failureDetails.GetInnerFailure()),
+		IsNonRetriable: !failureDetails.GetRetriable(), // need to negate this
+	}
+}
+
+func ConvertTaskFailureDetails(taskFailureDetails *protos.TaskFailureDetails) *dtmbprotos.FailureDetails {
+	if taskFailureDetails == nil {
+		return nil
+	}
+
+	return &dtmbprotos.FailureDetails{
+		ErrorType:    taskFailureDetails.GetErrorType(),
+		ErrorMessage: taskFailureDetails.GetErrorMessage(),
+		StackTrace:   taskFailureDetails.GetStackTrace().GetValue(),
+		InnerFailure: ConvertTaskFailureDetails(taskFailureDetails.GetInnerFailure()),
+		Retriable:    !taskFailureDetails.GetIsNonRetriable(),
+	}
+}
+
+func ConvertEvent(event *dtmbprotos.Event) (*protos.HistoryEvent, error) {
+	switch typedEvent := event.GetEventType().(type) {
+	case *dtmbprotos.Event_ExecutionStarted:
+		return &protos.HistoryEvent{
+			EventId:   int32(event.GetSequenceNumber()), // is this correct?
+			Timestamp: event.GetTimestamp(),
+			EventType: &protos.HistoryEvent_ExecutionStarted{
+				ExecutionStarted: &protos.ExecutionStartedEvent{
+					Name:    typedEvent.ExecutionStarted.GetName(),
+					Version: &wrapperspb.StringValue{Value: typedEvent.ExecutionStarted.GetVersion()},
+					Input:   &wrapperspb.StringValue{Value: string(typedEvent.ExecutionStarted.GetInput())}, // what if this is not string
```

Comment:
DTFx SDKs support strings only, although DTMB supports generic binary data. As long as you use DTFx only, it will work

Reply:
@cgillum should the protos in Durable Task be updates to support either bytes or string for Input / Output? Right now the protos can only accept strings and this seem problematic unless you want me to base64 encode all input, base64 decode the result etc

Reply:
No, all SDKs and engines exclusively use strings for data and won’t be able to benefit from bytes. Just convert them to/from UTF-8 bytes when calling Ale’s APIs. That or we should change the APIs to accept strings.

Reply:
Converting from a UTF8 string to a byte slice is always possible and safe. But storing binary (i.e. non-UTF8) data in a string requires encoding to formats like Base64. Unless we have a reason for saying otherwise, I'd prefer the backend to work with byte slices since they are opaque payloads anyways. Clients can interpret the byte slices as UTF8 strings if they wish.

Reply:
Yes the Go documentation is correct. But in this case, if we know that all current SDKs only use strings, then you can be sure that the data that is in that `byte[]` is a UTF8-encoded string. So the conversion `byte[] -> string` is safe.

ASCII standard (0-127) is a subset of UTF8, so there's no issue there.

Reply:
Yes agreed.